packages/hash/api/src/db/postgres/link/updateLink.ts

```diff
@@ -0,0 +1,133 @@
+import { Connection } from "../types";
+import { DbClient, DbLink, DbLinkVersion } from "../../adapter";
+import { getEntityLatestVersion } from "../entity";
+import { DbEntityNotFoundError } from "../..";
+import { genId } from "../../../util";
+import { getIndexedLinks } from "./sql/links.util";
+import { requireTransaction } from "../util";
+import { getLink } from "./getLink";
+import { DbLinkNotFoundError } from "../../errors";
+import {
+  insertLinkVersionRow,
+  updateLinkVersionRow,
+  updateLinkVersionIndices,
+} from "./sql/link_versions.util";
+
+export const updateLink = async (
+  existingConnection: Connection,
+  params: Parameters<DbClient["updateLink"]>[0],
+): Promise<DbLink> =>
+  requireTransaction(existingConnection)(async (conn) => {
+    const { sourceAccountId, linkId, updatedByAccountId } = params;
+
+    const previousDbLink = await getLink(conn, { sourceAccountId, linkId });
+
+    if (!previousDbLink) {
+      throw new DbLinkNotFoundError(params);
+    }
+
+    const now = new Date();
+
+    const { sourceEntityId, index: previousIndex, path } = previousDbLink;
+
+    if (previousIndex === undefined) {
+      throw new Error(
+        "Cannot update index of link that previously didn't have an index",
+      );
+    }
```

Comment:
How might we remove an index from a link if we want to change an array of links from ordered to unordered?

I'm wondering if we might want to check for null or undefined instead (`!= null`).

We could just address this if we ever need to.

Reply:
> I'm wondering if we might want to check for null or undefined instead (!= null).

Because of the implementation of `mapDbRowsToDbLink`, `dbLink.index` will never be `null` (only `undefined` or `number`) so think the `previousIndex === undefined` conditional can stay.

> How might we remove an index from a link if we want to change an array of links from ordered to unordered?

Interesting question, my instinct is that whether or not a collection of outgoing links with the same `path` are ordered or not should be specified in the schema where the link is defined.

Up until now I've considered the `index` to be an extension of the `path`, and because we currently disallow users from updating the `path` of a link and require them to re-create a link if they wanted to do that, I think we should do the same if a user wanted to change whether a certain link is indexed or not.

Reply:
Makes sense, thank you